src/utils.ts

```diff
@@ -0,0 +1,564 @@
+        symbol: token.symbol as string,
+      };
+
+      return parsedToken;
+    }),
+    raw: tx.raw as string,
+  };
+
+  return parsedTx;
+};
+
+/**
+ * Downloads a block from the full_node using the `block_at_height` API
+ *
+ * @param txId - The block txId
+ * @param noCache - Prevents downloading the block from cache as a reorg may have ocurred
+ */
+export const getBlockByTxId = async (txId: string, noCache: boolean = false) => {
+  return downloadTx(txId, noCache);
+};
+
+/**
+ * Syncs to the latest block
+ *
+ * @generator
+ * @yields {StatusEvent} A status event indicating if a block at a height was successfully sent, \
+ * if an error ocurred or if a reorg ocurred.
+ */
+export async function* syncToLatestBlock(): AsyncGenerator<StatusEvent> {
+  const ourBestBlock: Block = await getWalletServiceBestBlock();
+  const fullNodeBestBlock: Block = await getFullNodeBestBlock();
+
+  // Check if our best block is still in the fullnode's chain
+  const ourBestBlockInFullNode = await getBlockByTxId(ourBestBlock.txId, true);
+
+  if (!ourBestBlockInFullNode.success) {
+    yield {
+      type: 'error',
+      success: false,
+      message: 'Could not validate our best block',
+      error: ourBestBlockInFullNode.message,
+    };
+
+    return;
+  }
+
+  const { meta } = ourBestBlockInFullNode;
+
+  if ((meta.voided_by &&
+       meta.voided_by.length &&
+         meta.voided_by.length > 0)) {
+    yield {
+      type: 'reorg',
+      success: false,
+      message: 'Our best block was voided, we should reorg.',
+    };
+
+    return;
+  }
+
+  if (ourBestBlock.height > meta.height) {
+    yield {
+      type: 'reorg',
+      success: false,
+      message: 'Our height is higher than the wallet-service\'s height, we should reorg.',
+    };
+
+    return;
+  }
+
+  console.log('Best block is valid.');
+  let success = true;
+
+  blockLoop:
+    for (let i = ourBestBlock.height + 1; i <= fullNodeBestBlock.height; i++) {
+    const block: FullBlock = await downloadBlockByHeight(i);
+    const preparedBlock: PreparedTx = prepareTx(block);
+
+    // Ignore parents[0] because it is a block
+    const blockTxs = [
+      block.parents[1],
+      block.parents[2],
+    ];
+
+    // Download block transactions
+    const txs: FullTx[] = await recursivelyDownloadTx(block.txId, blockTxs);
+
+    // We will send the block only after all transactions were downloaded
+    // to be sure that all downloads were succesfull since there is no
+    // ROLLBACK yet on the wallet-service.
+    const sendBlockResponse: ApiResponse = await sendTx(preparedBlock);
+
+    if (!sendBlockResponse.success) {
+      console.log(sendBlockResponse);
+      yield {
+        type: 'error',
+        success: false,
+        message: `Failure on block ${preparedBlock.tx_id}`,
+      };
+
+      success = false;
+
+      break;
+    }
+
+    // Exclude duplicates:
+    const uniqueTxs: FullTx[] = txs.reduce((acc: FullTx[], tx: FullTx): FullTx[] => {
```

Comment:
O(n^2)? I feel we can do it using a dictionary in O(n). It can get really slow if a block has too many transactions.

Reply:
I agree. I left this to be refactored later and forgot about it, sorry

Done in f654dbe6c0f0dce363b8348c33be758f05df87b5

Reply:
Great. Thanks!